src/computation/1_computation_parallel.cpp

```diff
@@ -327,3 +241,462 @@ void ComputationParallel::applyBoundaryValuesTop()
         discretization_->g(i, discretization_->vJEnd() -1) = discretization_->v(i, discretization_->vJEnd() -1);
     }
 }
+
+void ComputationParallel::uvExchangeHorizontal()
+{
+    // the +4 in nCells+4 is from the consistent number of halo-cells in the staggered grid
+
+    // the even processes: send left, receive left, send right, receive right
+    if ((partitioning.ownRankNo() % 2) == 0)
+    {
+        // left
+        if (partitioning_.ownPartitionContainsLeftBoundary())
+        {
+            applyBoundaryValuesLeft();
+        } else 
+        {
+            // u
+            std::array<double, discretization_->nCells()[1]+4> u_to_other_ghost_left;
+            for (int j = -2; j < discretization_->nCells()[1] +2; j++)
```

Comment:
for the for loops use
```
std::array<double, discretization_->uJEnd - discretization_->uJBegin> u_to_other_ghost_left;
int j = discretization_->uJBegin < discretization_->uJEnd
```
instead of `int j = -2; j < discretization_->nCells()[1] +2` to make sure you always iterate over the complete boundaries, indepent of boundary/halo cells as they may differ

Reply:
This is a design choice I made: I think by sending the whole slice we are more independent of the different sizes of the  boundary/halo cells and can less easily err by sending too much or too little